File: java/src/main/java/org/digma/intellij/plugin/idea/psi/java/EndpointDiscovery.java
package org.digma.intellij.plugin.idea.psi.java;

import com.intellij.openapi.progress.EmptyProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Computable;
import com.intellij.psi.PsiFile;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.search.SearchScope;
import org.digma.intellij.plugin.model.discovery.DocumentInfo;
import org.digma.intellij.plugin.model.discovery.EndpointInfo;
import org.digma.intellij.plugin.model.discovery.MethodInfo;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;
import java.util.function.Supplier;

public interface EndpointDiscovery {

    //must run with read access
    //using searchScope supplier because building SearchScope needs read access
    List<EndpointInfo> lookForEndpoints(@NotNull Supplier<SearchScope> searchScopeSupplier);

    // default method uses fileScope. however, in some cases logic could be bit different
    default List<EndpointInfo> lookForEndpoints(@NotNull PsiFile psiFile) {
        return lookForEndpoints(() -> GlobalSearchScope.fileScope(psiFile));
    }

    default void endpointDiscovery(@NotNull PsiFile psiFile, @NotNull DocumentInfo documentInfo) {
        List<EndpointInfo> endpointInfos = lookForEndpoints(psiFile);

        for (EndpointInfo endpointInfo : endpointInfos) {
            var methodId = endpointInfo.getContainingMethodId();
            final MethodInfo methodInfo = documentInfo.getMethods().get(methodId);
            //this method must exist in the document info
            Objects.requireNonNull(methodInfo, "method info " + methodId + " must exist in DocumentInfo for " + documentInfo.getFileUri());

            methodInfo.addEndpoint(endpointInfo);
        }
    }


    default void runInReadAccess(Project project, Runnable runnable) {
        ProgressManager.getInstance().runProcess(() -> DumbService.getInstance(project).runReadActionInSmartMode(runnable), new EmptyProgressIndicator());
    }

    default <T> T runInReadAccessWithResult(Project project, Computable<T> computable) {
        return ProgressManager.getInstance().runProcess(() -> DumbService.getInstance(project).runReadActionInSmartMode(computable), new EmptyProgressIndicator());
    }

}

File: jvm-common/src/main/kotlin/org/digma/intellij/plugin/idea/discovery/endpoint/EndpointDiscovery.kt
package org.digma.intellij.plugin.idea.discovery.endpoint

import com.intellij.psi.PsiFile
import com.intellij.psi.search.GlobalSearchScope
import org.digma.intellij.plugin.common.SearchScopeProvider
import org.digma.intellij.plugin.errorreporting.ErrorReporter
import org.digma.intellij.plugin.model.discovery.DocumentInfo
import org.digma.intellij.plugin.model.discovery.EndpointInfo
import java.util.Objects

abstract class EndpointDiscovery {

    abstract fun getName(): String

    abstract suspend fun lookForEndpoints(searchScopeProvider: SearchScopeProvider): List<EndpointInfo>?


    suspend fun endpointDiscovery(psiFile: PsiFile, documentInfo: DocumentInfo) {

        val endpointInfos = lookForEndpoints { GlobalSearchScope.fileScope(psiFile) }

        endpointInfos?.let { infos ->
            for (endpointInfo in infos) {
                val methodId = endpointInfo.containingMethodId
                val methodInfo = documentInfo.methods[methodId]

                try {
                    //this method must exist in the document info
                    Objects.requireNonNull(methodInfo) {
                        val methodIds = documentInfo.methods.map { it.key }.joinToString(",")
                        "method info $methodId must exist in DocumentInfo for ${documentInfo.fileUri} , methods: [$methodIds]"
                    }
                    methodInfo?.addEndpoint(endpointInfo)
                } catch (e: Exception) {
                    //report the exception and continue instead of failing the whole discovery.
                    ErrorReporter.getInstance().reportError("${this::class.java}.endpointDiscovery.${getName()}", e)
                }
            }
        }
    }
}